//package antlr
//
//// The root of the ANTLR exception hierarchy. In general, ANTLR tracks just
////  3 kinds of errors: prediction errors, failed predicate errors, and
////  mismatched input errors. In each case, the parser knows where it is
////  in the input, where it is in the ATN, the rule invocation stack,
////  and what kind of problem occurred.
//
//pub trait RecognitionException {
//  GetOffendingToken() Token
//  GetMessage() &str
//  GetInputStream() IntStream
//}
//
//pub struct BaseRecognitionException {
//  message:        &str
//  recognizer:     Recognizer
//  offending_token: Token
//  offending_state: i32
//  ctx:            RuleContext
//  input:          IntStream
//}
//
//impl BaseRecognitionException {§//  fn new(&self, message &str, recognizer Recognizer, input: IntStream, ctx: RuleContext) -> *BaseRecognitionException {
//
//  // todo
//  //  Error.call(this)
//  //
//  //  if (!!Error.capture_stack_trace) {
//  //        Error.capture_stack_trace(this, RecognitionException)
//  //  } else {
//  //    let stack = NewError().stack;
//  //  }
//  // TODO may be able to use - "runtime" func Stack(buf []byte, all bool) i32
//
//  let t = new(BaseRecognitionException);
//
//  self.message = message
//  self.recognizer = recognizer
//  self.input = input
//  self.ctx = ctx
//  // The current {@link Token} when an error occurred. Since not all streams
//  // support accessing symbols by index, we have to track the {@link Token}
//  // instance itself.
//  self.offending_token = nil
//  // Get the ATN state number the parser was in at the time the error
//  // occurred. For {@link NoViableAltException} and
//  // {@link LexerNoViableAltException} exceptions, this is the
//  // {@link DecisionState} number. For others, it is the state whose outgoing
//  // edge we couldn't Match.
//  self.offending_state = -1
//  if self.recognizer != nil {
//    self.offending_state = self.recognizer.GetState()
//  }
//
//  return t
//}
//
//pub fn GetMessage(&self, ) -> &str {
//  return self.message
//}
//
//pub fn GetOffendingToken(&self, ) -> Token {
//  return self.offending_token
//}
//
//pub fn GetInputStream(&self, ) -> IntStream {
//  return self.input
//}
//
//// <p>If the state number is not known, b method returns -1.</p>
//
////
//// Gets the set of input symbols which could potentially follow the
//// previously Matched symbol at the time b exception was panicn.
////
//// <p>If the set of expected tokens is not known and could not be computed,
//// b method returns {@code nil}.</p>
////
//// @return The set of token types that could potentially follow the current
//// state in the ATN, or {@code nil} if the information is not available.
//// /
//fn get_expected_tokens(&self, ) -> *IntervalSet {
//  if self.recognizer != nil {
//    return self.recognizer.GetATN().get_expected_tokens(self.offending_state, self.ctx)
//  }
//
//  return nil
//}
//
//pub fn String(&self, ) -> &str {
//  return self.message
//}
//
//pub struct LexerNoViableAltException {
//  *BaseRecognitionException
//
//  start_index:     i32
//  dead_end_configs: ATNConfigSet
//}
//
//impl LexerNoViableAltException {§//  fn new(&self, lexer Lexer, input CharStream, start_index: i32, dead_end_configs: ATNConfigSet) -> *LexerNoViableAltException {
//
//  let l = new(LexerNoViableAltException);
//
//  self.BaseRecognitionException = BaseRecognitionException::new("", lexer, input, nil)
//
//  self.start_index = start_index
//  self.dead_end_configs = dead_end_configs
//
//  return l
//}
//
//pub fn String(&self, ) -> &str {
//  let symbol = "";
//  if self.start_index >= 0 && self.start_index < self.input.Size() {
//    symbol = self.input.(CharStream).GetTextFromInterval(NewInterval(self.start_index, self.start_index))
//  }
//  return "LexerNoViableAltException" + symbol
//}
//
//pub struct NoViableAltException {
//  *BaseRecognitionException
//
//  start_token:     Token
//  offending_token: Token
//  ctx:            ParserRuleContext
//  dead_end_configs: ATNConfigSet
//}
//
//// Indicates that the parser could not decide which of two or more paths
//// to take based upon the remaining input. It tracks the starting token
//// of the offending input and also knows where the parser was
//// in the various paths when the error. Reported by ReportNoViableAlternative()
////
//impl NoViableAltException {§//  fn new(&self, recognizer Parser, input TokenStream, start_token Token, offending_token Token, dead_end_configs: ATNConfigSet, ctx: ParserRuleContext) -> *NoViableAltException {
//
//  if ctx == nil {
//    ctx = recognizer.GetParserRuleContext()
//  }
//
//  if offending_token == nil {
//    offending_token = recognizer.GetCurrentToken()
//  }
//
//  if start_token == nil {
//    start_token = recognizer.GetCurrentToken()
//  }
//
//  if input == nil {
//    input = recognizer.GetInputStream().(TokenStream)
//  }
//
//  let n = new(NoViableAltException);
//  self.BaseRecognitionException = BaseRecognitionException::new("", recognizer, input, ctx)
//
//  // Which configurations did we try at input.Index() that couldn't Match
//  // input.LT(1)?//
//  self.dead_end_configs = dead_end_configs
//  // The token object at the start index the input stream might
//  // not be buffering tokens so get a reference to it. (At the
//  // time the error occurred, of course the stream needs to keep a
//  // buffer all of the tokens but later we might not have access to those.)
//  self.start_token = start_token
//  self.offending_token = offending_token
//
//  return n
//}
//
//pub struct InputMisMatchException {
//  *BaseRecognitionException
//}
//
//// This signifies any kind of mismatched input exceptions such as
//// when the current input does not Match the expected token.
////
//impl InputMisMatchException {§//  fn new(&self, recognizer: Parser) -> *InputMisMatchException {
//
//  let i = new(InputMisMatchException);
//  self.BaseRecognitionException = BaseRecognitionException::new("", recognizer, recognizer.GetInputStream(), recognizer.GetParserRuleContext())
//
//  self.offending_token = recognizer.GetCurrentToken()
//
//  return i
//
//}
//
//// A semantic predicate failed during validation. Validation of predicates
//// occurs when normally parsing the alternative just like Matching a token.
//// Disambiguating predicate evaluation occurs when we test a predicate during
//// prediction.
//
//pub struct FailedPredicateException {
//  *BaseRecognitionException
//
//  rule_index:      i32
//  predicate_index: i32
//  predicate:      &str
//}
//
//impl FailedPredicateException {§//  fn new(&self, recognizer Parser, predicate: &str, message: &str) -> *FailedPredicateException {
//
//  let f = new(FailedPredicateException);
//
//  self.BaseRecognitionException = BaseRecognitionException::new(self.format_message(predicate, message), recognizer, recognizer.GetInputStream(), recognizer.GetParserRuleContext())
//
//  let s = recognizer.GetInterpreter().atn.states[recognizer.GetState()];
//  let trans = self.GetTransitions()[0];
//  if trans2, let ok = trans.(*PredicateTransition); ok {;
//    self.rule_index = trans2.rule_index
//    self.predicate_index = trans2.pred_index
//  } else {
//    self.rule_index = 0
//    self.predicate_index = 0
//  }
//  self.predicate = predicate
//  self.offending_token = recognizer.GetCurrentToken()
//
//  return f
//}
//
//fn format_message(&self, predicate, message: &str) -> &str {
//  if message != "" {
//    return message
//  }
//
//  return "failed predicate: {" + predicate + "}?"
//}
//
//pub struct ParseCancellationException {
//}
//
//impl ParseCancellationException {§//  fn new(&self, ) -> *ParseCancellationException {
//  //  Error.call(this)
//  //  Error.capture_stack_trace(this, ParseCancellationException)
//  return new(ParseCancellationException)
//}
